Fixes #266 - problem with reactimate, adds dswitch to dunai

miguel-negrao opened the pull request:
Example code that caused error before this commit:

```Haskell
import Data.IORef
import FRP.Yampa

main = do
  ref <- newIORef (0.0::Double)
  reactimate senseI (senseRest ref) (actuate ref) sf

senseI = do
  putStrLn "SenseI"
  return 45

senseRest ref _ = do
  v <- readIORef ref
  putStrLn "SenseRest"
  return (0.1, Nothing)

actuate ref _ x = do
  v <- readIORef ref
  putStrLn $ "actuate " ++ show x
  writeIORef ref (v+1)
  return (v >= 1)

sf = identity
```

miguel-negrao commented on dunai/src/Control/Monad/Trans/MSF/Except.hs:
```diff
@@ -273,6 +273,14 @@ switch sf f = catchS ef  f
            (b,me)  <- liftTransS sf -< a
            inExceptT -< ExceptT $ return $ maybe (Right b) Left me
 
+-- * Analog to Yampa's dswitch, with Maybe instead of Event
```
I believe this is just comment, not haddock, right ? I copy-pasted from switch.

miguel-negrao replied:
Changed to haddock.

ivanperez-keera commented:
An issue has just been closed to address the fact that the haddock syntax in `switch` was the section syntax, not the function documentation syntax.

That's independent from this issues.

@miguel-negrao Could you please rebase this on top of the current `develop`? It will give you a conflict, but it will also reduce the size of the commit (by 1-2 lines).

Also, please see my questions regarding the syntax (two are requests for types, the other is a syntax question).

miguel-negrao commented:
@ivanperez-keera I can't seem to find your questions regarding the syntax. Are they comments in this PR ?

I've rebased on developed. I've also simplified switch's implementation using `throwMaybe`. I believe the code becomes clearer with this change, but check if you feel the same.

ivanperez-keera commented:
Thanks!

I started it as a review of your code. The questions were attached to the implementation:

![Screenshot from 2022-03-23 11-10-07](https://user-images.githubusercontent.com/1162320/159731752-3b04fb58-4f0b-45c0-98e1-c929e4b4a07b.png)

miguel-negrao commented:
I think this happened to me once. I think you have to hit something like "submit review" for the review to go live and become available.

ivanperez-keera commented:
It's ok. You've answered the question, so that's all I needed. Thanks!

ivanperez-keera commented:
I cannot replicate the issue. I've done some work on this issue and left a comment here: https://github.com/ivanperez-keera/dunai/issues/266#issuecomment-1086673235

ivanperez-keera commented:
This PR will be closed but some changes have been merged and @miguel-negrao credited for them.

See https://github.com/ivanperez-keera/dunai/issues/266#issuecomment-1086705215 for comments regrading the regression test and #276 for a follow-up issue on the simplification of switch.

Thanks!
